src/chmatch.c

```diff
@@ -37,45 +40,111 @@ SEXP chmatch(SEXP x, SEXP table, R_len_t nomatch, Rboolean in) {
+    // used to be called chmatch2 before v1.12.2 and was in rbindlist.c. New implementation from 1.12.2 here in chmatch.c
+    // if nuniq==tablelen then there are no dups and the simpler chmatch branch below happens instead to avoid these allocations etc
+    // see end of file for benchmark
+    //                                                                                        uniq         dups
+    // For example: A,B,C,B,D,E,A,A   =>   A(TL=1),B(2),C(3),D(4),E(5)   =>   dupMap    1  2  3  5  6 | 8  7  4
+    //                                                                        dupLink   7  8          |    6     (blank=0)
+    int *counts = (int *)calloc(nuniq, sizeof(int));
+    int *map =    (int *)calloc(tablelen+nuniq, sizeof(int));  // +nuniq to store a 0 at the end of each group
+    if (!counts || !map) {
+      // # nocov start
+      for (int i=0; i<tablelen; i++) SET_TRUELENGTH(td[i], 0);
+      savetl_end();
+      error("Failed to allocate %lld bytes working memory in chmatchdup: length(table)=%d length(unique(table))=%d", (tablelen*2+nuniq)*sizeof(int), tablelen, nuniq);
+      // # nocov end
+    }
+    for (int i=0; i<tablelen; ++i) counts[-TRUELENGTH(td[i])-1]++;
+    for (int i=0, sum=0; i<nuniq; ++i) { int tt=counts[i]; counts[i]=sum; sum+=tt+1; }
+    for (int i=0; i<tablelen; ++i) map[counts[-TRUELENGTH(td[i])-1]++] = i+1;           // 0 is left ending each group thanks to the calloc
+    for (int i=0, last=0; i<nuniq; ++i) {int tt=counts[i]+1; counts[i]=last; last=tt;}  // rewind counts to the beginning of each group
+    for (int i=0; i<xlen; ++i) {
+      int u = TRUELENGTH(xd[i]);
+      if (u<0) {
+        int w = counts[-u-1]++;
+        if (map[w]) { ansd[i]=map[w]; continue; }
+        SET_TRUELENGTH(xd[i],0); // w falls on ending 0 marker: dups used up; any more dups should return nomatch
+        // we still need the 0-setting loop at the end of this function because often there will be some values in table that are not matched to at all.
+      }
+      ansd[i] = nomatch;
+    }
+    free(counts);
+    free(map);
+  } else if (chin) {
+    for (int i=0; i<xlen; i++) {
+      ansd[i] = TRUELENGTH(xd[i])<0;
     }
   } else {
-    for (i=0; i<length(x); i++) {
-      m = TRUELENGTH(STRING_ELT(x,i));
-      INTEGER(ans)[i] = (m<0) ? -m : nomatch;
+    for (int i=0; i<xlen; i++) {
+      int m = TRUELENGTH(xd[i]);
+      ansd[i] = (m<0) ? -m : nomatch;
     }
   }
-  for (i=0; i<length(table); i++)
-    SET_TRUELENGTH(STRING_ELT(table,i),0);  // reinstate 0 rather than leave the -i-1
+  for (int i=0; i<tablelen; i++)
+    SET_TRUELENGTH(td[i], 0);  // reinstate 0 rather than leave the -i-1
   savetl_end();
   UNPROTECT(1);
   return(ans);
 }
 
-SEXP chmatchwrapper(SEXP x, SEXP table, SEXP nomatch, SEXP in) {
-  return(chmatch(x,table,INTEGER(nomatch)[0],LOGICAL(in)[0]));
+// for internal use from C :
+SEXP chmatch(SEXP x, SEXP table, int nomatch) {  // chin=  chmatchdup=
```

Comment:
if it is for internal use from C why not return C object type instead of SEXP?

Reply:
Just because that SEXP is most often a vector. SEXP is convenient even internally so we can leave R to do the memory management for us.  Otherwise for single values, yes C type internally.